src/systems/filecoin_markets/storage_market/storage_market_actor_code.go

```diff
@@ -92,248 +101,319 @@ func (a *StorageMarketActorCode_I) WithdrawBalance(rt Runtime, amountRequested a
 	//     - insufficient balance lock up
 	// this operation will be unrolled if any of the above triggers a throw
-	for i, newDeal := range newStorageDeals {
-		if st._validateNewStorageDeal(rt, newDeal) {
-			st._lockFundsForStorageDeal(rt, newDeal)
-			id := st._generateStorageDealID(rt, newDeal)
-
-			onchainDeal := &deal.OnChainDeal_I{
-				ID_:               id,
-				Deal_:             newDeal,
-				LastPaymentEpoch_: block.ChainEpoch(LastPaymentEpochNone), // -1 = inactive
-			}
-			st.Deals()[id] = onchainDeal
-			response[i] = PublishStorageDealSuccess(id)
-		} else {
-			response[i] = PublishStorageDealError()
+	for _, newDeal := range newStorageDeals {
+		p := newDeal.Proposal()
+
+		st._rtAbortIfNewDealInvalid(rt, newDeal)
+
+		// Before any operations that check the balance tables for funds, execute all deferred
+		// deal state updates.
+		//
+		// Note: as an optimization, implementations may cache efficient data structures indicating
+		// which of the following set of updates are redundant and can be skipped.
+		st._rtUpdatePendingDealStatesForParty(rt, p.Client())
+		st._rtUpdatePendingDealStatesForParty(rt, p.Provider())
+
+		st._rtLockBalanceUntrusted(rt, p.Client(), p.ClientBalanceRequirement())
+		st._rtLockBalanceUntrusted(rt, p.Provider(), p.ProviderBalanceRequirement())
+
+		id := st._generateStorageDealID(rt, newDeal)
+
+		onchainDeal := &deal.OnChainDeal_I{
+			ID_:               id,
+			Deal_:             newDeal,
+			SectorStartEpoch_: block.ChainEpoch_None,
 		}
+		st.Deals()[id] = onchainDeal
 	}
 
 	UpdateRelease(rt, h, st)
-
-	return response
 }
 
-func (a *StorageMarketActorCode_I) VerifyPublishedDealIDs(rt Runtime, dealIDs []deal.DealID) {
-
-	TODO() // verify StorageMinerActor
+func (a *StorageMarketActorCode_I) VerifyDealsOnSectorPreCommit(rt Runtime, dealIDs deal.DealIDs, sectorInfo sector.SectorPreCommitInfo) {
+	rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_StorageMiner)
+	minerAddr := rt.ImmediateCaller()
 
 	h, st := a.State(rt)
 
-	for _, dealID := range dealIDs {
-
-		publishedDeal := st._safeGetOnChainDeal(rt, dealID)
-		st._assertPublishedDealState(rt, dealID)
-
-		dealP := publishedDeal.Deal().Proposal()
-		st._assertDealStartAfterCurrEpoch(rt, dealP)
+	for _, dealID := range dealIDs.Items() {
+		deal, _ := st._rtGetOnChainDealOrAbort(rt, dealID)
+		_rtAbortIfDealInvalidForNewSectorSeal(rt, minerAddr, sectorInfo.Expiration(), deal)
 
 		// deal must not expire before the maximum allowable epoch between pre and prove commits
 		// we do not have to check if the deal has expired at ProveCommit
 		// if the MAX_PROVE_COMMIT_SECTOR_EPOCH constraint is not violated
-		st._assertDealExpireAfterMaxProveCommitWindow(rt, dealP)
-
+		// if dealP.EndEpoch() <= (rt.CurrEpoch() + sector.MAX_PROVE_COMMIT_SECTOR_EPOCH) {
+		// 	rt.AbortStateMsg("Deal might expire before prove commit.")
+		// }
+		TODO() // TODO: confirm that the above is no longer required given the new deal expiration logic
```

Comment:
Not sure if deal expiration logic matters here but I think we can remove these lines. They are saying if the deals are so short that they might expire before prove commit, we should not accept them in PreCommit. What this PR is doing is to check `_rtAbortIfDealInvalidForNewSectorSeal` at both PreCommit and ProveCommit which will reject deals whose StartEpoch has elapsed which is a stronger requirement that what we previously had (we allowed deals to Start in between PreCommit and ProveCommit). I am supportive of this change 👍